verilog/analysis/checkers/dff_name_style_rule.h

```diff
@@ -0,0 +1,134 @@
+// Copyright 2017-2023 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VERIBLE_VERILOG_ANALYSIS_CHECKERS_CONSTRAINT_NAME_STYLE_RULE_H_
+#define VERIBLE_VERILOG_ANALYSIS_CHECKERS_CONSTRAINT_NAME_STYLE_RULE_H_
+
+#include <optional>
+#include <set>
+#include <string>
+
+#include "common/analysis/lint_rule_status.h"
+#include "common/analysis/syntax_tree_lint_rule.h"
+#include "common/text/symbol.h"
+#include "common/text/syntax_tree_context.h"
+#include "verilog/analysis/descriptions.h"
+
+namespace verilog {
+namespace analysis {
+
+// DffNameStyleRule checks that D flip flops use appropriate naming conventions
+// for both their input and outputs.
+//
+// The rule can be configured via specifying a comma-separated list of
+// suffixes (one for input, one for output). Providing an empty list
+// means no checks for the corresponding field
+//
+// For example, the defaults are equivalent to:
+//  +dff-name-style=output:reg,r,ff,q;input:next,n,d
+//  Which gets expanded into
+//  `valid_output_suffixes` = { "reg", "r", "ff", "q" }
+//  `valid_input_suffixes` = { "next", "n", "d" }
+//
+// Given a nonblocking assignment inside an always_ff block
+// we will then check the left-hand side (lhs) and right
+// hand side (rhs) of the assignment against the valid
+// suffixes
+//
+// Given "data_q <= data_n", we will check that:
+//   - data_q ends with any of { "_reg", "_r", "_ff", "_q" }
+//   - data_n ends with any of { "_next", "_n", "_d" }
+//
+// If those checks succeed, we will also check that their prefixes
+// are equal.
+//  data_q <= data_n -> OK
+//  data_q <= something_n -> WRONG: "data" != "something"
+//
+// The right-hand side also accepts constant integers:
+//  data_q  <= 0 -> OK
+//  data_q3 <= 0 -> OK
+//
+// An extension of this rule could perform checks outside
+// always_ff blocks, so that
+//  assign data_q = something
+//  always_comb data_q = something
+//  data_d <= something
+// produce warnings
+//
+// Following [1] we also allow for using trailing numbers
+// in the identifiers to specify the pipeline stage where
+// a given signal comes from.
+//
+// Under this convention, data_q3 should be driven by the
+// previous stage data_q2
+//
```

Comment:
What about cases where LHS is a compound expression like:
```systemverilog
{zz_q2, yy_q2} = ...

(It's ok to declare this as a TODO.)
```

Reply:
This would fail currently due to

https://github.com/IEncinas10/verible/blob/9ed09223b6df913140cbc0f4f3aac345ff4600cf/verilog/analysis/checkers/dff_name_style_rule.cc#L174

that was basically added to ease the check. It is not clear to me how to deal with all the possibilities consistently.

Maybe I could add a comment saying something like: "Currently this rule only allows for simple identifiers, but it could be expanded to handle more complex references"